### DISCUSSION TITLE
Broker tests, with rust integration tests

### DISCUSSION THREAD

**Discussion by seam345** [bytebeamio/rumqtt#445] [General]
To learn MQTT spec and better myself at rust tests I have been converting the MQTT spec to rust integration tests https://gitlab.com/seam345/learning-mqtt

In doing so I have started to come across some cases where RUMQTTD *appears* to fail the spec (currently [MQTT-3.1.2-3] in Version 3.1.1) failed cargo test https://gitlab.com/seam345/learning-mqtt/-/jobs/2980022830 test code https://gitlab.com/seam345/learning-mqtt/-/blob/352172c00c6c02a2642e185181ae00c13874e501/tests/connect_packet.rs#L328.

Opened as a discussion instead of PR for a few reasons:
1. I haven't looked into RUMQTT testing strategy properly yet.
2. I have just started making these tests and so am not 100% confident in the validity of them, I would like to spend some more time confirming they are correct before opening a PR for a bug.
3. Once I am happy with the tests validity how would you like me to report bugs (if I find any)?
    1. A PR with the failing tests added to this repo
    2. A PR outlining the failed spec and a detailed description for how to reproduce
    4. Another way? 
4. Would RUMQTT be interested in this style integration test in general? Once finished getting through the spec I can submit as a batch

**Comment by tekjar** [comment]
We would be interested @seam345 . Infact, we are writing tests [here](https://github.com/bytebeamio/mqttwrk/tree/conformance-test). Not complete though. There's also a new version of broker coming [soon](https://github.com/bytebeamio/rumqtt/tree/rumqtt-ng)
